src/EditorFeatures/Core/Implementation/FindReferences/AbstractFindReferencesService.cs

```diff
@@ -51,36 +59,127 @@ protected AbstractFindReferencesService(IEnumerable<IReferencedSymbolsPresenter>
+            // TODO: Merging references from SymbolFinder and external providers might lead to duplicate or counter-intuitive results.
+            // TODO: For now, we avoid merging and just display the results either from SymbolFinder or the external result providers but not both.
+            if (builder.Count == 0)
+            {
+                await AddSymbolReferencesAsync(document, position, builder, waitContext).ConfigureAwait(false);
+            }
 
             // realize the list here so that the consumer await'ing the result doesn't lazily cause
             // them to be created on an inappropriate thread.
-            return q.ToList();
+            return builder.ToArrayAndFree();
         }
 
         public bool TryFindReferences(Document document, int position, IWaitContext waitContext)
         {
             var cancellationToken = waitContext.CancellationToken;
+            var workspace = document.Project.Solution.Workspace;
+
+            // First see if we have any external navigable item references.
+            // If so, we display the results as navigable items.
+            var succeeded = TryFindNavigableItemsReferencesAsync(document, position, waitContext).WaitAndGetResult(cancellationToken);            
+            if (succeeded)
+            {
+                return true;
+            }
 
+            // Otherwise, fall back to displaying SymbolFinder based references.
             var result = this.FindReferencedSymbolsAsync(document, position, waitContext).WaitAndGetResult(cancellationToken);
+            return DisplayReferences(result);
+        }
+
+        /// <summary>
+        /// Attempts to finds and display navigable item references, including the references provided by external providers.
+        /// </summary>
+        /// <returns>False if there are no external references or display was not successful.</returns>
+        private async Task<bool> TryFindNavigableItemsReferencesAsync(Document document, int position, IWaitContext waitContext)
+        {
+            if (_externalReferencesProviders.Any())
+            {
+                var cancellationToken = waitContext.CancellationToken;
+                var builder = ArrayBuilder<INavigableItem>.GetInstance();
+                await AddExternalReferencesAsync(document, position, builder, cancellationToken).ConfigureAwait(false);
+
+                // TODO: Merging references from SymbolFinder and external providers might lead to duplicate or counter-intuitive results.
+                // TODO: For now, we avoid merging and just display the results either from SymbolFinder or the external result providers but not both.
+                if (builder.Count > 0 && DisplayReferences(builder))
+                {
+                    builder.Free();
+                    return true;
+                }
+
+                builder.Free();
```

Comment:
please use some sort of 'using' construct so we don't have to remember to free on all codepaths.